server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java

```diff
@@ -86,63 +96,103 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
-
-            for (var translogOp : request.changes()) {
-                final var op = withPrimaryTerm(translogOp, primary.getOperationPrimaryTerm());
+        private void performOnPrimary(IndexShard primary, Request request, List<Engine.Result> accumulator, int offset, Consumer<List<Engine.Result>> result, Consumer<Exception> failure) {
+            for (int itemsToProcess = offset; itemsToProcess < request.changes().size(); itemsToProcess++) {
+                final var op = request.changes().get(itemsToProcess);
+                final var opWithPrimary = withPrimaryTerm(op, primary.getOperationPrimaryTerm());
                 if (primary.getMaxSeqNoOfUpdatesOrDeletes() < request.maxSeqNoOfUpdatesOrDeletes()) {
                     primary.advanceMaxSeqNoOfUpdatesOrDeletes(request.maxSeqNoOfUpdatesOrDeletes());
                 }
-                var result = primary.applyTranslogOperation(op, Engine.Operation.Origin.PRIMARY);
-                if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                    throw new RuntimeException("Mapping update required");
+                Engine.Result engineResult = null;
+                try {
+                    engineResult = primary.applyTranslogOperation(opWithPrimary, Engine.Operation.Origin.PRIMARY);
+                } catch (IOException e) {
+                    failure.accept(e);
+                    return;
+                }
+                if (engineResult.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED ||
+                    engineResult.getResultType() == Engine.Result.Type.FAILURE) {
+                    // Retry to process the translog operations, once the mappings are updated
+                    retryWhenMappingsAreUpdated(primary, request, accumulator, itemsToProcess, result, failure);
+                    return;
                 }
+                accumulator.add(engineResult);
+            }
+            result.accept(accumulator);
+        }
 
-                location = syncOperationResultOrThrow(result, location);
+        private void retryWhenMappingsAreUpdated(IndexShard primary, Request request,List<Engine.Result> accumulator, int position, Consumer<List<Engine.Result>> result, Consumer<Exception> failure) {
+            var indexName = request.shardId().getIndexName();
+            var currentMappingVersion = clusterService.state().metadata().index(indexName).getMappingVersion();
+            var clusterStateObserver = new ClusterStateObserver(clusterService, LOGGER);
+            clusterStateObserver.waitForNextChange(
+                new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        performOnPrimary(primary, request, accumulator, position, result, failure);
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                    }
```

Comment:
Shouldn't this trigger the listener or so?

Reply:
I thought this would be used to free resources and close whatever needs to closed.

Reply:
I guess maybe throwing an error makes sense here.